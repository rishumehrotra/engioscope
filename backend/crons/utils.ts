import { CronJob } from 'cron';
import cronTime from 'cron-time-generator';
import debug from 'debug';

const cronLog = debug('cron');

export const runJob = (name: string, when: (time: typeof cronTime) => string, onTick: () => Promise<void>) => {
  const timePattern = when(cronTime);

  cronLog('Setting up cron for ', name, 'with pattern', timePattern);
  const j = new CronJob(
    timePattern,
    async () => {
      cronLog('Starting cron for', name);
      try {
        await onTick();
        cronLog('Done cron for', name);
      } catch (error) {
        cronLog('Cron for ', name, 'FAILED', error);
      }
    }
  );
  j.start();
};
