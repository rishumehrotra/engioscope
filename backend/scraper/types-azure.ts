import type workItemIconSvgs from '../work-item-icon-svgs';

export type ProjectState = 'all' | 'createPending' | 'deleted' | 'deleting' | 'new' | 'unchanged' | 'wellFormed';
export type ProjectVisibility = 'private' | 'public';
export type DefinitionTypeStatus = 'disabled' | 'enabled' | 'paused';
export type DefinitionType = 'build' | 'xaml';
export type BuildStatus = 'all' | 'cancelling' | 'completed' | 'inProgress' | 'none' | 'notStarted' | 'postponed';
export type BuildResult = 'canceled' | 'failed' | 'none' | 'partiallySucceeded' | 'succeeded';
export type QueuePriority = 'aboveNormal' | 'belowNormal' | 'high' | 'low' | 'normal';
export type BuildReason = 'all' | 'batchedCI' | 'buildCompletion' | 'checkInShelveset'
| 'individualCI' | 'manual' | 'none' | 'pullRequest' | 'schedule' | 'scheduleForced'
| 'triggered' | 'userCreated' | 'validateShelveset';
export type ReleaseStatus = 'abandoned' | 'active' | 'draft' | 'undefined';
export type ReleaseReason = 'continuousIntegration' | 'manual' | 'none' | 'pullRequest' | 'schedule';
export type EnvironmentStatus = 'canceled' | 'inProgress' | 'notStarted'
| 'partiallySucceeded' | 'queued' | 'rejected' | 'scheduled' | 'succeeded' | 'undefined';
export type DeploymentReason = 'automated' | 'manual' | 'none' | 'redeployTrigger' | 'scheduled';
export type DeploymentStatus = 'all' | 'failed' | 'inProgress' | 'notDeployed' | 'partiallySucceeded' | 'succeeded' | 'undefined';
export type DeploymentOperationStatus = 'All' | 'Approved' | 'Canceled' | 'Cancelling'
| 'Deferred' | 'EvaluatingGates' | 'GateFailed' | 'ManualInterventionPending'
| 'Pending' | 'PhaseCanceled' | 'PhaseFailed' | 'PhaseInProgress'
| 'PhasePartiallySucceeded' | 'PhaseSucceeded' | 'Queued' | 'QueuedForAgent'
| 'QueuedForPipeline' | 'Rejected' | 'Scheduled' | 'Undefined';
export type PullRequestStatus = 'abandoned' | 'active' | 'all' | 'completed' | 'notSet';
export type PullRequestAsyncStatus = 'conflicts' | 'failure' | 'notSet' | 'queued' | 'rejectedByPolicy' | 'succeeded';
export type GitPullRequestMergeStrategy = 'noFastForward' | 'rebase' | 'rebaseMerge' | 'squash';

export type TeamProjectReference = {
  id: string;
  name: string;
  description?: string;
  url?: string;
  state: ProjectState;
  revision?: number;
  visibility: ProjectVisibility;
  lastUpdatedTime: Date;
};

export type DefinitionReference = {
  createdDate?: Date;
  id: number;
  name: string;
  path: string;
  project: TeamProjectReference;
  queueStatus: DefinitionTypeStatus;
  revision: number;
  type: DefinitionType;
  uri: string;
  url: string;
};

export type IdentityRef = {
  displayName: string;
  url?: string;
  id: string;
  uniqueName?: string;
  imageUrl?: string;
  descriptor?: string;
};

export type AgentPoolQueue = {
  id: number;
  name: string;
  pool: {
    id: number;
    name: string;
  };
};

export type GitRepository = {
  id: string;
  name: string;
  url: string;
  defaultBranch?: string;
  size: number;
  remoteUrl: string;
  sshUrl: string;
  webUrl: string;
  project: TeamProjectReference;
};

export type BuildRepository = {
  id: string;
  type: string;
  name: string;
  url: string;
  clean: null;
  checkoutSubmodules: boolean;
};

export type BuildDefinitionReference = DefinitionReference & {
  latestBuild?: Build;
  latestCompletedBuild?: Build;
  process: {
    type: 1 | 2;
  };
};

export type Build = {
  plans: { planId: string }[];
  id: number;
  buildNumber: string;
  status: BuildStatus;
  result: BuildResult;
  queueTime: Date;
  startTime: Date;
  finishTime: Date;
  url: string;
  definition: DefinitionReference;
  buildNumberRevision: number;
  project: TeamProjectReference;
  uri: string;
  sourceBranch: string;
  sourceVersion: string;
  queue: AgentPoolQueue;
  priority: QueuePriority;
  reason: BuildReason;
  requestedFor: IdentityRef;
  requestedBy: IdentityRef;
  lastChangeDate: Date;
  lastChangedBy: IdentityRef;
  parameters: string;
  orchestrationPlan: { planId: string };
  repository: BuildRepository;
  keepForever: boolean;
  retainedByRelease: boolean;
  triggeredByBuild: Build | null;
};

export type DeploymentAttempt = {
  id: number;
  deploymentId: number;
  attempt: number;
  reason: DeploymentReason;
  status: DeploymentStatus;
  operationStatus: DeploymentOperationStatus;
  // releaseDeploymentPhases: ReleaseDeployPhase[]
  requestedBy: IdentityRef;
  requestedFor: IdentityRef;
  queuedOn: Date;
  lastModifiedBy: IdentityRef;
  lastModifiedOn: Date;
  hasStarted: boolean;
};

export type ReleaseCondition = {
  conditionType: 'artifact' | 'environmentState' | 'event' | 'undefined';
  name: string;
  result: boolean;
  value: string;
};

export type ReleaseEnvironment = {
  id: number;
  releaseId: number;
  name: string;
  status: EnvironmentStatus;
  deploySteps: DeploymentAttempt[];
  rank: number;
  definitionEnvironmentId: number;
  conditions: ReleaseCondition[];
  // incomplete
};

export type ArtifactSourceReference = {
  id: string;
  name: string | null;
};

export type Artifact = {
  sourceId: string;
  type: string;
  alias: string;
  definitionReference: {
    artifactSourceDefinitionUrl: ArtifactSourceReference;
    buildUri: ArtifactSourceReference;
    definition: ArtifactSourceReference;
    pullRequestSourceBranchCommitId: ArtifactSourceReference;
    pullRequestId: ArtifactSourceReference;
    pullRequestIterationId: ArtifactSourceReference;
    pullRequestMergeCommitId: ArtifactSourceReference;
    project: ArtifactSourceReference;
    pullRequestSourceBranch: ArtifactSourceReference;
    pullRequestTargetBranch: ArtifactSourceReference;
    repository: ArtifactSourceReference;
    requestedFor: ArtifactSourceReference;
    requestedForId: ArtifactSourceReference;
    branch: ArtifactSourceReference;
    [x: string]: ArtifactSourceReference;
  };
};

export type ProjectReference = {
  id: string;
  name: string;
};

export type ReleaseDefinitionShallowReference = {
  id: number;
  name: string;
  path: string;
  projectReference: ProjectReference | null;
  url: string;
};

export type Release = {
  id: number;
  name: string;
  status: ReleaseStatus;
  createdOn: Date;
  modifiedOn: Date;
  modifiedBy: IdentityRef;
  createdBy: IdentityRef;
  environments: ReleaseEnvironment[];
  artifacts: Artifact[];
  releaseDefinition: ReleaseDefinitionShallowReference;
  releaseDefinitionRevision: number;
  description: string;
  reason: ReleaseReason;
  releaseNameFormat: string;
  keepForever: boolean;
  definitionSnapshotRevision: number;
  logsContainerUrl: string;
  url: string;
  projectReference: { id: string; name: string };
};

export type GitUserDate = {
  date: Date;
  email: string;
  imageUrl: string;
  name: string;
};

export type GitCommitRef = {
  commitId: string;
  url: string;
  author: GitUserDate;
  committer: GitUserDate;
  comment: string;
  changeCounts: {
    Add: number;
    Edit: number;
    Delete: number;
  };
};

export type GitPullRequestCompletionOptions = {
  mergeCommitMessage: string;
  mergeStrategy: GitPullRequestMergeStrategy;
  bypassReason: string;
  triggeredByAutoComplete: boolean;
};

export type GitPullRequest = {
  repository: {
    id: string;
    name: string;
    url: string;
    project: TeamProjectReference;
  };
  pullRequestId: number;
  codeReviewId: number;
  status: PullRequestStatus;
  createdBy: IdentityRef;
  creationDate: Date;
  closedDate: Date;
  title: string;
  description: string;
  sourceRefName: string;
  targetRefName: string;
  mergeStatus: PullRequestAsyncStatus;
  isDraft: boolean;
  mergeId: string;
  lastMergeSourceCommit: GitCommitRef;
  lastMergeTargetCommit: GitCommitRef;
  lastMergeCommit: GitCommitRef;
  reviewers: (IdentityRef & { vote: number })[];
  url: string;
  completionOptions: GitPullRequestCompletionOptions;
  supportsIterations: boolean;
  completionQueueTime: Date;
};

export type GitBranchStats = {
  aheadCount: number;
  behindCount: number;
  commit: GitCommitRef;
  isBaseVersion: boolean;
  name: string;
  url: string;
};

export type ReleaseReference = {
  id: number;
  name: string;
  environmentId: number;
  environmentName: string | null;
  definitionId: number;
  environmentDefinitionId: number;
  environmentDefinitionName: string | null;
  creationDate: Date;
};

export type TestRun = {
  id: number;
  name: string;
  url: string;
  build: { id: string };
  isAutomated: boolean;
  owner: IdentityRef;
  project: { id: string; name: string };
  startedDate: Date;
  completedDate: Date;
  state: 'Unspecified' |'NotStarted' | 'InProgress' | 'Completed' | 'Waiting' | 'Aborted' | 'NeedsInvestigation';
  totalTests: number;
  incompleteTests: number;
  notApplicableTests: number;
  passedTests: number;
  unanalyzedTests: number;
  revision: number;
  release: ReleaseReference;
  webAccessUrl: string;
};

export type CodeCoverageData = {
  buildFlavor: string;
  buildPlatform: string;
  coverageStats: {
    covered: number;
    delta?: number;
    isDeltaAvailable?: boolean;
    label: string;
    position: number;
    total: number;
  }[];
};

export type CodeCoverageSummary = {
  coverageData: CodeCoverageData[];
  build: {
    id: string;
    url: string;
  };
};

export type ReleaseDefinitionEnvironment = {
  id: number;
  name: string;
  rank: number;
  owner: IdentityRef;
  variableGroups: [];
  schedules: [];
  // currentRelease: {
  //   id: number,
  //   url: string,
  //   _links: {}
  // },
  retentionPolicy: {
    daysToKeep: number;
    releasesToKeep: number;
    retainBuild: boolean;
  };
  // properties: {},
  preDeploymentGates: {
    id: number;
    gatesOptions: null;
    gates: [];
  };
  postDeploymentGates: {
    id: number;
    gatesOptions: null;
    gates: [];
  };
  environmentTriggers: [];
  badgeUrl: string;
  conditions: ReleaseCondition[];
};

export type ReleaseDefinition = {
  source: 'ibiza' | 'portalExtensionApi' | 'restApi' | 'undefined' | 'userInterface';
  revision: number;
  description: string;
  createdBy: IdentityRef;
  createdOn: Date;
  modifiedBy: IdentityRef;
  modifiedOn: Date;
  isDeleted: boolean;
  variableGroups: null;
  environments: ReleaseDefinitionEnvironment[];
  releaseNameFormat: string;
  retentionPolicy: { daysToKeep: number };
  // properties: {},
  id: number;
  name: string;
  path: string;
  projectReference: null;
  url: string;
};

export type WorkItemTypeReference = {
  name: string;
  url: string;
};

export type WorkItemTypeCategory = {
  name: string;
  referenceName: string;
  defaultWorkItemType: WorkItemTypeReference;
  workItemTypes: WorkItemTypeReference[];
  url: string;
};

export type WorkItemQueryFlatResult = {
  queryType: 'flat';
  workItems: {
    id: number;
    url: string;
  }[];
};

export type WorkItemQueryHierarchialResult = {
  queryType: 'oneHop' | 'tree';
  workItemRelations: {
    rel: string | null;
    source: null | { id: number; url: string };
    target: null | { id: number; url: string };
  }[];
};

export type WorkItemQueryResult<T extends WorkItemQueryFlatResult | WorkItemQueryHierarchialResult> = {
  asOf: Date;
  columns: {
    name: string;
    referenceName: string;
    url: string;
  }[];
} & T;

export type WorkItemTypeFieldInstance = {
  defaultValue: string | null;
  helpText?: string;
  alwaysRequired: boolean;
  referenceName: string;
  name: string;
  url: string;
};

export type WorkItemType = {
  name: string;
  referenceName: string;
  description: string;
  color: string;
  icon: { id: keyof typeof workItemIconSvgs; url: string };
  isDisabled: boolean;
  xmlForm: string;
  fields: WorkItemTypeFieldInstance[];
  fieldInstances: WorkItemTypeFieldInstance[];
  transitions: Record<string, { to: string }[]>;
  states: { name: string; color: string; category: string }[];
  url: string;
  rootCause?: string[];
};

export type WorkItem = {
  id: number;
  rev: number;
  fields: {
    'System.TeamProject': string;
    'System.WorkItemType': string;
    'System.State': string;
    'System.ChangedDate': Date;
    'System.CreatedDate': Date;
    'System.Title': string;
    'System.Description': string;
    'Microsoft.VSTS.TCM.AutomationStatus'?: 'Not Automated' | 'Automated';
    'Microsoft.VSTS.Common.ClosedDate'?: Date;
    'Microsoft.VSTS.Common.StateChangeDate'?: Date;
    'Microsoft.VSTS.Common.Priority'?: number;
    'Microsoft.VSTS.Common.Severity'?: string;
  } & Record<string, string>;
  url: string;
};

export type WorkItemLink = {
  rel: 'System.LinkTypes.Hierarchy'
  | 'System.LinkTypes.Related'
  | 'Microsoft.VSTS.Common.TestedBy'
  | 'System.LinkTypes.Duplicate'
  | 'Microsoft.VSTS.TestCase.SharedStepReferencedBy'
  | 'Microsoft.VSTS.Common.Affects'
  | 'System.LinkTypes.Dependency'
  | 'Microsoft.VSTS.TestCase.SharedParameterReferencedBy';
  attributes: {
    linkType: string;
    sourceId: number;
    targetId: number;
    isActive: boolean;
    changedDate: Date;
    changedBy: {
      id: string;
      displayName: string;
      uniqueName: string;
      descriptor: string;
    };
    comment: string | null;
    changedOperation: 'remove' | 'create';
    sourceProjectId: string;
    targetProjectId: string;
  };
};

export type WorkItemRevision = {
  id: number;
  rev: number;
  fields: {
    'System.TeamProject': string;
    'System.WorkItemType': string;
    'System.State': string;
    'System.ChangedDate': Date;
    'System.ChangedBy': string;
    'System.CreatedDate': Date;
    'System.CreatedBy': string;
    'System.AuthorizedDate': Date;
    'System.RevisedDate': Date;
    'System.Rev': number;
    'System.PersonId': number;
    'System.Title': string;
  };
};

export type WorkItemField = {
  name: string;
  referenceName: string;
  url: string;
  description: string | null;
  type: 'boolean' | 'dateTime' | 'double' | 'guid' | 'history' | 'html'
  | 'identity' | 'integer' | 'picklistDouble' | 'picklistInteger'
  | 'picklistString' | 'plainText' | 'string' | 'treePath';
  isDeleted: boolean;
  isIdentity: boolean;
  isPicklist: boolean;
  isPicklistSuggested: boolean;
  isQueryable: boolean;
  picklistId: string;
  readOnly: boolean;
  supportedOperations: { name: string; referenceName: string }[];
  usage: 'none' | 'tree' | 'workItem' | 'workItemLink' | 'workItemTypeExtension';
};

type PolicyBase = {
  createdBy: IdentityRef;
  createdDate: Date;
  id: number;
  isBlocking: boolean;
  isDeleted: boolean;
  isEnabled: boolean;
};

type BranchPolicyType<TypeName extends string, AdditionalSettings> = PolicyBase & {
  settings: {
    scope: {
      refName: string;
      matchKind: 'Exact';
      repositoryId: string;
    }[];
  } & AdditionalSettings;
  type: {
    id: string;
    url: string;
    displayName: TypeName;
  };
};

type RepoPolicyType<TypeName extends string, AdditionalSettings> = PolicyBase & {
  settings: {
    scope: {
      repositoryId: string;
    }[];
  } & AdditionalSettings;
  type: {
    id: string;
    url: string;
    displayName: TypeName;
  };
};

type FileSizeRestrictionPolicy = RepoPolicyType<'File size restriction', {
  maximumGitBlobSizeInBytes: number;
  useUncompressedSize: boolean;
}>;
type PathLengthRestrictionPolicy = RepoPolicyType<'Path Length restriction', { maxPathLength: number }>;
type ReservedNamesRestrictionPolicy = RepoPolicyType<'Reserved names restriction', Record<string, never>>;
type MinimumNumberOfReviewersPolicy = BranchPolicyType<'Minimum number of reviewers', {
  minimumApproverCount: number;
  creatorVoteCounts: boolean;
  allowDownvotes: boolean;
  resetOnSourcePush: boolean;
}>;
type CommentRequirementsPolicy = BranchPolicyType<'Comment requirements', Record<string, never>>;
type WorkItemLinkingPolicy = BranchPolicyType<'Work item linking', Record<string, never>>;
type BuildPolicy = BranchPolicyType<'Build', {
  buildDefinitionId: number;
  queueOnSourceUpdateOnly: boolean;
  manualQueueOnly: boolean;
  displayName: string | null;
  validDuration: number;
}>;
type RequiredReviewersPolicy = BranchPolicyType<'Required reviewers', {
  requiredReviewerIds: string[];
  filenamePatterns?: string[];
}>;
type RequireMergeStrategyPolicy = BranchPolicyType<'Require a merge strategy', {
  allowRebase?: boolean;
}>;

export type PolicyConfiguration =
  | FileSizeRestrictionPolicy
  | PathLengthRestrictionPolicy
  | ReservedNamesRestrictionPolicy
  | MinimumNumberOfReviewersPolicy
  | CommentRequirementsPolicy
  | WorkItemLinkingPolicy
  | BuildPolicy
  | RequiredReviewersPolicy
  | RequireMergeStrategyPolicy;

export const isFileSizeRestrictionPolicy = (policy: PolicyConfiguration): policy is FileSizeRestrictionPolicy => (
  policy.type.displayName === 'File size restriction'
);
export const isPathLengthRestrictionPolicy = (policy: PolicyConfiguration): policy is PathLengthRestrictionPolicy => (
  policy.type.displayName === 'Path Length restriction'
);
export const isReservedNamesRestrictionPolicy = (policy: PolicyConfiguration): policy is ReservedNamesRestrictionPolicy => (
  policy.type.displayName === 'Reserved names restriction'
);
export const isMinimumNumberOfReviewersPolicy = (policy: PolicyConfiguration): policy is MinimumNumberOfReviewersPolicy => (
  policy.type.displayName === 'Minimum number of reviewers'
);
export const isCommentRequirementsPolicy = (policy: PolicyConfiguration): policy is CommentRequirementsPolicy => (
  policy.type.displayName === 'Comment requirements'
);
export const isWorkItemLinkingPolicy = (policy: PolicyConfiguration): policy is WorkItemLinkingPolicy => (
  policy.type.displayName === 'Work item linking'
);
export const isBuildPolicy = (policy: PolicyConfiguration): policy is BuildPolicy => (
  policy.type.displayName === 'Build'
);
export const isRequiredReviewersPolicy = (policy: PolicyConfiguration): policy is RequiredReviewersPolicy => (
  policy.type.displayName === 'Required reviewers'
);
export const isRequireMergeStrategyPolicy = (policy: PolicyConfiguration): policy is RequireMergeStrategyPolicy => (
  policy.type.displayName === 'Require a merge strategy'
);
