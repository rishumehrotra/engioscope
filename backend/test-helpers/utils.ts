export const generateRandomTestId = () => {
  return Math.floor(Math.max() * Math.random()).toString(36);
};
