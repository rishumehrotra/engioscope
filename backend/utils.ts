// Not sure why the /// <referennce ... /> is needed, but tsc fails without it
// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="../node_modules/@types/node/fs.d.ts" />

import { promises as fs } from 'fs';
import ms from 'ms';
import { allPass, compose, not } from 'rambda';
import type { ParsedConfig } from './scraper/parse-config.js';

export const pastDate = (past?: string) => {
  if (!past) return new Date();

  const d = new Date();
  d.setMilliseconds(d.getMilliseconds() - ms(past));
  return d;
};

export const isAfter = (time: string) => (date: Date) => (
  date.getTime() > pastDate(time).getTime()
);
export const isWithin = (time: string) => (date: Date) => (
  date.getTime() <= pastDate(time).getTime()
);

export const isNewerThan = (date1: Date) => (date2: Date) => (
  date2.getTime() > date1.getTime()
);

export const shortDateFormat = (date: Date) => [
  date.toLocaleString('default', { month: 'short' }),
  date.getDate()
].join(' ');

export const isMaster = (branchName: string) => [
  'refs/heads/master', 'refs/heads/main'
].includes(branchName);

export const normalizeBranchName = (branchName: string) => branchName.replace('refs/heads/', '');

export const doesFileExist = async (filePath: string) => {
  try {
    await fs.access(filePath);
    return true;
  } catch (e) {
    return false;
  }
};

export const range = (num: number) => [...Array(num).keys()];

export const chunkArray = <T>(array: T[], chunkSize: number) => (
  range(Math.ceil(array.length / chunkSize))
    .map(i => array.slice(i * chunkSize, (i + 1) * chunkSize))
);

export const unique = <T>(xs: T[]) => [...new Set(xs)];

const weekNumbers = [12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1];

export const weekLimits = weekNumbers
  .map(weekNumber => [
    pastDate(`${weekNumber * 7} days`),
    pastDate(`${(weekNumber - 1) * 7} days`)
  ] as const);

export const weeks = weekNumbers
  .map(weekNumber => allPass([
    isAfter(`${weekNumber * 7} days`),
    compose(not, isAfter(`${(weekNumber - 1) * 7} days`))
  ]));

type SettleSeriesType<T> = {
  status: 'fulfilled';
  value: T;
} | {
  status: 'rejected';
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  reason: any;
};

export const mapSettleSeries = <T, U>(xs: T[], fn: (x: T) => Promise<U>) => (
  xs.reduce<Promise<SettleSeriesType<U>[]>>((acc, x) => (
    acc.then(accArr => (
      fn(x)
        .then(value => ({ status: 'fulfilled', value } as const))
        .catch(err => ({ status: 'rejected', reason: err } as const))
        .then(result => [...accArr, result])
    ))
  ), Promise.resolve([]))
);

export const startTimer = () => {
  const start = Date.now();
  return () => `${(Date.now() - start) / 1000}s`;
};

export const queryPeriodDays = (config: ParsedConfig) => Math.round(
  (Date.now() - config.azure.queryFrom.getTime()) / (1000 * 60 * 60 * 24)
);

export const wait = (ms: number) => new Promise(resolve => { setTimeout(resolve, ms); });

export const retry = <T>(fn: () => Promise<T>, { retryCount = 10, waitTime = 1 } = {}): Promise<T> => (
  fn().catch(async err => {
    if (retryCount <= 0) throw err;
    await wait(waitTime * 1000);
    return retry(fn, { retryCount: retryCount - 1, waitTime });
  })
);
