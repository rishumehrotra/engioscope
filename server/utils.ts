import ms from 'ms';
import { promises as fs } from 'fs';

const oneMinute = 1000 * 60;
const oneHour = oneMinute * 60;

export const pastDate = (past?: string) => {
  if (!past) return new Date();

  const d = new Date();
  d.setMilliseconds(d.getMilliseconds() - ms(past));
  return d;
};

const isWithin = (time: string) => (date: Date) => date > pastDate(time);
export const isWithinFortnight = isWithin('15 days');

/* eslint-disable @typescript-eslint/indent */
export const statsStrings = (emptyValue: string, transform: (a: number) => string):
  [range: (a: number[]) => string, average: (a: number[]) => string] => {
  const rangeAsString = (arr: number[]) => {
    switch (arr.length) {
      case 0: return emptyValue;
      case 1: return transform(arr[0]);
      default: return `${transform(Math.min(...arr))} - ${transform(Math.max(...arr))}`;
    }
  };

  const averageAsString = (arr: number[]) => {
    if (arr.length === 0) return emptyValue;
    return transform(arr.reduce((a, b) => a + b, 0) / arr.length);
  };

  return [rangeAsString, averageAsString];
};
/* eslint-enable @typescript-eslint/indent */

export const minutes = (ms: number) => {
  const m = Math.ceil(ms / oneMinute);
  if (m <= 1) return '1 m';
  return `${m} m`;
};

export const hours = (ms: number) => {
  const h = Math.round(ms / oneHour);
  if (h === 1) return '1 h';
  return `${h.toFixed(1)} h`;
};

export const shortDateFormat = (date: Date) => [
  date.toLocaleString('default', { month: 'short' }),
  date.getDate()
].join(' ');

export const isMaster = (branchName: string) => [
  'refs/heads/master', 'refs/heads/main'
].includes(branchName);

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
export const assertDefined = <T>(x: T | undefined) => x!;

export const divideBy = (divisor: number) => (dividend: number) => dividend / divisor;
export const getFirst = <T>(list: T[]): T | undefined => list[0];
export const filter = <T>(fn: (x: T) => boolean) => (xs: T[]) => xs.filter(fn);
export const map = <T, R>(fn: (x: T) => R) => (
  (xs: T[]) => xs.map(fn)
);

export const doesFileExist = async (filePath: string) => {
  try {
    await fs.access(filePath);
    return true;
  } catch (e) {
    return false;
  }
};
