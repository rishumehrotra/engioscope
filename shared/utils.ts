import { add } from 'rambda';
import { maybe } from './maybe.js';

export const oneSecondInMs = 1000;
export const oneMinuteInMs = 60 * oneSecondInMs;
export const oneHourInMs = 60 * oneMinuteInMs;
export const oneDayInMs = 24 * oneHourInMs;
export const oneFortnightInMs = 15 * oneDayInMs;
export const oneMonthInMs = 30 * oneDayInMs;
export const oneYearInMs = 365 * oneDayInMs;

export const exists = <T>(x: T | undefined | null): x is T => (
  x !== null && x !== undefined
);

export const divide = (numerator: number, denominator: number) => (
  maybe(numerator / denominator)
);

export const toPercentage = (value: number) => `${Math.round(value * 100)}%`;

export const combineColumnsInArray = <T>(combiner: (a: T, b: T) => T) => (rows: T[][]) => (
  rows.reduce<T[]>((acc, row) => {
    row.forEach((val, index) => {
      acc[index] = combiner(acc[index], val);
    });
    return acc;
  }, [])
);

export const addColumnsInArray = combineColumnsInArray<number>((a, b) => add(a || 0, b));

export const mapObj = <T, U>(xform: (x: T) => U) => (obj: Record<string, T>) => (
  Object.fromEntries(
    Object.entries(obj)
      .map(([key, value]) => [key, xform(value)])
  )
);

export const merge = <T>(obj1: T, obj2: T) => ({ ...obj1, ...obj2 });
