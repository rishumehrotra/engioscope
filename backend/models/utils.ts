import { z } from 'zod';
import { oneWeekInMs } from '../../shared/utils.js';

export const queryContextInputParser = z.tuple([
  z.string(),
  z.string(),
  z.date(),
  z.date(),
]);

export type QueryContext = z.infer<typeof queryContextInputParser>;

export const fromContext = (x: QueryContext) => {
  return {
    collectionName: x[0],
    project: x[1],
    startDate: x[2],
    endDate: x[3],
  };
};

export const durationIndex = (startDate: Date, field: string, durationMs: number) => ({
  $trunc: { $divide: [{ $subtract: [field, startDate] }, durationMs] },
});

export const weekIndexValue = (startDate: Date, field: string) =>
  durationIndex(startDate, field, oneWeekInMs);
