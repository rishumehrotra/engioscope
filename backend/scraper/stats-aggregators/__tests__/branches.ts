import { GitBranchStats } from '../../types-azure';
import { pastDate } from '../../../utils';
import aggregateBranches from '../branches';

test('aggregate-branches with a single old branch behind trunk', () => {
  expect(aggregateBranches([{
    aheadCount: 0,
    behindCount: 1,
    commit: { committer: { date: pastDate('25 days') } }
  }] as GitBranchStats[])).toMatchSnapshot();
});
