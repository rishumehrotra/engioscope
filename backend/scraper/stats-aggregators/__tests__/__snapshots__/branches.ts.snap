// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aggregate-branches with a single old branch behind trunk 1`] = `
Object {
  "count": 1,
  "indicators": Array [
    Object {
      "name": "Total",
      "tooltip": "Total numbers of branches in the repository",
      "value": 1,
    },
    Object {
      "name": "Active",
      "tooltip": "Active development branches which are in-sync with master",
      "value": 0,
    },
    Object {
      "name": "Abandoned",
      "tooltip": "Inactive development branches which are out-of-sync with master, but contain commits which are not present on master",
      "value": 0,
    },
    Object {
      "name": "Delete candidates",
      "tooltip": "Inactive development branches which are in-sync with master",
      "value": 0,
    },
    Object {
      "name": "Possibly conflicting",
      "tooltip": "Branches that are significantly out of sync with master",
      "value": 0,
    },
  ],
  "name": "Branches",
}
`;
